File: leetcodeQuestion34/leetcodeQuestion34.py
from typing import List


class Solution:
    def firstOccurrence(self, nums: List[int], target: int) -> int:
        start = 0
        end = len(nums) - 1
        leftPos = -1
        while start <= end:
            mid = start + ((end - start) // 2)
            if target == nums[mid]:
                leftPos = mid
                end = mid - 1
            elif target < nums[mid]:
                end = mid - 1
            elif target > nums[mid]:
                start = mid + 1
        return leftPos

    def lastOccurrence(self, nums: List[int], target: int):
        start = 0
        end = len(nums) - 1
        rightPos = -1
        while start <= end:
            mid = start + ((end - start) // 2)
            if target == nums[mid]:
                rightPos = mid
                start = mid + 1
            elif target < nums[mid]:
                end = mid - 1
            elif target > nums[mid]:
                start = mid + 1
        return rightPos

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        result: List[int] = []
        result.append(self.firstOccurrence(nums, target))
        result.append(self.lastOccurrence(nums, target))
        return result

    def __init__(self, nums: List[int], target: int) -> None:
        print(self.searchRange(nums, target))


arr = [1, 2, 3, 3, 3, 4, 7]
Solution(arr, 1)

File: leetcodeQuestion34/leetcodeQuestion34.cpp
#include <bits/stdc++.h>
using namespace std;

int firstOccurrence(int arr[], int n, int target)
{
    int start = 0;
    int end = n - 1;
    int mid = (start + (end - start) / 2);
    int leftPos = -1;
    while (start <= end)
    {
        if (arr[mid] == target)
        {
            leftPos = mid;
            end = mid - 1;
        }
        else if (arr[mid] < target)
        {
            start = mid + 1;
        }
        else if (arr[mid] > target)
        {
            end = mid - 1;
        }
        mid = (start + (end - start) / 2);
    }
    return leftPos;
}

int lastOccurrence(int arr[], int n, int target)
{
    int start = 0;
    int end = n - 1;
    int mid = (start + (end - start) / 2);
    int rightPos = -1;
    while (start <= end)
    {
        if (arr[mid] == target)
        {
            rightPos = mid;
            start = mid + 1;
        }
        else if (arr[mid] < target)
        {
            start = mid + 1;
        }
        else if (arr[mid] > target)
        {
            end = mid - 1;
        }
        mid = (start + (end - start) / 2);
    }
    return rightPos;
}

int main()
{
    int arr[7] = {1, 2, 3, 3, 3, 4, 7};
    cout << "[" << firstOccurrence(arr, 7, 3) << ", " << lastOccurrence(arr, 7, 3) << "]" << endl;
}